### ISSUE TITLE
How is the actor KL divergence computed?

### ISSUE THREAD

**Issue by miriaford** [rraileanu/auto-drac#2]
Hi,

Thanks for open-sourcing this great work! I might have missed something simple, but could you please kindly explain why KL divergence is computed using only `- action_log_probs_aug.mean()` as in this line:

https://github.com/rraileanu/auto-drac/blob/master/ucb_rl2_meta/algo/drac.py#L95

If I understand correctly, the KL divergence should be between the action distribution computed from the original image and the action distribution computed from the augmented version. However only the augmented log_prob is considered here. Shouldn't the KL divergence include both (`q` is the augmented and `p` is the original): 

![image](https://user-images.githubusercontent.com/29965840/90423474-5f1ec600-e071-11ea-916d-0be6f4dc31aa.png)

**Comment by rraileanu** [comment]
Hi! Thanks for your interest, this is a great question! 

We only want to backprop gradients through the augmented policy and not through the original one (just like we do for the value function). Thus, we can disregard the loss component that depends on the original policy since its corresponding gradients are zero. 

Here is a more detailed derivation of the gradients. Hope this helps but let me know if you have other questions.

<img width="655" alt="kl_term" src="https://user-images.githubusercontent.com/55603802/90434029-b0a47080-e0c4-11ea-89b0-1131e7319b4c.png">

**Comment by miriaford** [comment]
I see, thank you so much for the explanation! It makes a lot more sense now. 

As a quick check for my understanding - this loss term is independent of the distribution type right? For example, if I want to use continuous action space with Gaussian parameterization, I can use the same `aug_loss = - action_log_probs_aug.mean()` (but computed by Gaussian)? 

Any other difference for continuous action space? Thanks again for your help!

**Comment by miriaford** [comment]
Thank you so much for your reply! 

I just checked wikipedia again, the KL divergence formula should be 
![image](https://user-images.githubusercontent.com/29965840/90456965-1a198480-e0af-11ea-88e3-1b9174f76166.png)

i.e. the first distribution should be the one being sampled from, since KL is asymmetrical. 

It looks like you are computing the loss of `KL( p || q_theta)` instead of `KL( q_theta || p)` ?

**Comment by rraileanu** [comment]
You are right, we are computing KL(p || q_theta) where p is the original distribution and q_theta is the augmented one. Sorry for the typo above.

**Comment by miriaford** [comment]
Should the paper also be corrected? Here's the arxiv version, section 3.2:

![image](https://user-images.githubusercontent.com/29965840/90528596-ab2c4200-e127-11ea-83ca-d0fc5f263487.png)

`q_theta` corresponds to the augmented policy, which should go to the second place in KL.

**Comment by rraileanu** [comment]
Yes, you are right. Thanks for pointing this out! This change will appear in the next arxiv update of the paper.
